extras/entropy/entropy.1
.TH ENTROPY 1 
.ad
.fi
.SH NAME
entropy
\-
compute data entropy
.SH SYNOPSIS
.na
.nf
.ad
.fi
\fBentropy\fR [\fB-b \fIblock_size\fR] \fIfile...\fR
.SH DESCRIPTION
.ad
.fi
\fBentropy\fR reports entropy values for the named files (standard
input by default). The entropy is computed with Shannon's formula
applied to byte streams. The result gives the average number of bits
of information per byte, under the assumption that the input has no
specific structure.

For each file, entropy is reported as follows:

.ti +4
file file_entropy block_count min_entropy max_entropy

Fields are tab separated.

.IP \fIfile\fR
The name of the file.
.IP \fIfile_entropy\fR
The entropy computed over the entire file contents.
.IP \fIblock_count\fR
The number of \fIblock_size\fR blocks used for the minimal and
maximal per-block entropy computation.
.IP "\fImin_entropy, max_entropy\fR"
The minimal and maximal per-block entropy for the named file.
No per-block entropy is computed for partial blocks.
.PP
Options:
.IP "\fB-b \fIblock_size\fR (default: 8192)"
The block size to be used for per-block entropies.
.SH BUGS
.ad
.fi
Entropy values can be misleading.
.IP \(bu
The per-file entropy computed over the entire file is larger
than any per-block entropy value for that same file.
.IP \(bu
The entropy values as reported by this program are valid only
for unstructured input. With real data such as text the result
is an overestimate.
.SH HISTORY
.na
.nf
.fi
.ad
This \fBentropy\fR command was written for the coroner's toolkit.
.SH LICENSE
.na
.nf
This software is distributed under the IBM Public License.
.SH SEE ALSO
.na
.nf
.ad
.fi
C.E.  Shannon, A mathematical theory of communication. Bell
System Technical Journal 27 (1948) 379-423, 632-656.
.nf
.na
http://cm.bell-labs.com/cm/ms/what/shannonday/shannon1948.ps.gz
.SH AUTHOR(S)
.na
.nf
Wietse Venema
IBM T.J. Watson Research
P.O. Box 704
Yorktown Heights, NY 10598, USA


CHANGES
Sat Aug 29 2009

o	grave-robber did not record the initial directory in the body file.

o	Bit-rot: FreeBSD 7 support, including ``struct acct'' changes.

o	Bit-rot: lastcomm format strings.

Sun Jan 11 20:01:27 EST 2009

o	Bit-rot: OpenBSD 4.x support (still TODO: file system tools).

Thu Jun  7 10:54:44 EDT 2007

o	Bit-rot: FreeBSD 6.x support.

Mon Apr  9 09:36:31 EDT 2007

o	Bit-rot: the 64-bit workaround for LINUX lseek() is no
	longer needed, and it didn't compile on some systems.

Thu Jan 11 16:10:29 EST 2007

o	Bugfix: the timeout command always reported zero exit status
	after voluntary child process exit.

Mon Jun 26 18:44:17 EDT 2006

o	Bit-rot: update #include statements because code either
	no longer compiled or compiled with warnings.

Fri Mar 17 09:25:57 EST 2006

o	Bit-rot: update #include statements because code either
	no longer compiled or compiled with warnings.

Tue Jan  6 12:25:32 EST 2004

o       Workaround: mactime treated "," as a special character in
	file names, now it's 0xff.

o	Support for ext[23]fs inode file sizes > 32bits.

o	Bugfix: ils file sizes larger than unsigned long.

Tue Oct 14 18:23:49 EDT 2003

o       pcat now fully supports FreeBSD systems that do not have
	/proc mounted (it's no longer mounted by default with
	FreeBSD 5). pcat tries to use /proc if it can, and uses
	ptrace() and gropes kernel memory if it has to.  Operation
	without /proc requires super-user privileges.

Thu Oct  2 09:14:00 EDT 2003

o       Completed support for UFS1 and UFS2 in FreeBSD 5.x. However,
	the pcat command still requires that the /proc file system
	is mounted. This will be fixed later.

Sat Aug 30 19:21:29 EDT 2003

o	Preliminary FreeBSD 5.0 port. UFS2 support is still to be done.

Fri Aug  1 10:07:39 EDT 2003

o	RedHat 9.0 fix by Florin Andrei for the file command.

Sat Oct  5 13:47:29 EDT 2002

o	Restored tctutils compatibility, which broke with 20020916.

Mon Sep 16 14:39:09 EDT 2002

o       The lowest inode examined is now the lowest possible inode
	number instead of ROOTINUM.

Fri Mar 22 14:41:32 EST 2002

o	Added support for OpenBSD 3.0.

Wed Dec 12 13:21:00 EST 2001

o	tm_print() and tm_string() did unnecessary %HEX expansion
	of $ and ^ characters (Philippe Bourgeois).

o       mactime mis-sorted dates before/after early September 2001
	(Philippe Bourgeois, Brian Carrier).

Thu Oct  4 15:48:42 EDT 2001

o       Linux can handle files > 2Gbyte, but only if we specify
	extra compiler flags.

o	Solaris 8/64bit needs an extra include file in pcat.c.

Sun Sep  9 17:23:13 EDT 2001

o       Discovered to my horror that unrm on ext2fs can have an
	off-by-one error when testing a bit flag in the data block
	allocation bitmaps - and this problem is dependent on the
	Linux version being used.

o       Made the timeout command a little more verbose so it is
	less of a mystery when the child process is killed.

Thu Sep  6 13:51:00 EDT 2001

o	New grave-robber -I option for better control of icat/pcat.

Mon May 21 10:11:19 EDT 2001

o       findkey command-line error check was too conservative,
	and disallowed threshold of 256 different byte values
	(Ovanes Manucharyan).

Mon May 14 13:11:54 EDT 2001

o	Patch 06 broke lib/pass.cache.pl, which broke mactime.

o	Patch 06 did not update the patchlevel file.

Fri Mar 16 14:25:11 EST 2001

o	Fixed rpm root directory when running on a corpse.
	Marco Walter (SUN).

o	Fixed file(1) output parsing problem with multiple :
	characters.  Marco Walter (SUN)

Thu Feb 22 11:16:40 EST 2001

o	Patch by Brian Carrier (Purdue) so that fs_read_random()
	properly maintains file offset information.

Thu Feb  8 10:40:02 EST 2001

o	grave-robber dies silently when no group file is found.
	This bites when examining a partial machine corpse
	(bin/grave-robber). Klaus-Peter Kossakowski.

o	The "do first looks" code did not relocate to $TCT_HOME
	(lib/tree.pl). Klaus-Peter Kossakowski.

Tue Jan 23 18:44:36 EST 2001

o       The grave-robber did not correctly handle multi-line df
	output records (ps_spy.pl, suck_free_inodes.pl).

o       The grave-robber mounted corpse inode scan looked only at
	the corpse's "root" file system (suck_free_inodes.pl).

Mon Jan 22 11:23:32 EST 2001

o       grave-robber enabled the default -lPO flags when -c was
	specified, causing the program to gripe and exit.
	
Tue Sep 26 10:55:24 EDT 2000

o       Fixed a bug in fs_read_block() that caused icat etc. to
	remember the wrong file seek position, so that icat etc.
	would call lseek() too often. The other possibility, not
	calling lseek() when it should be called, did not happen
	due to an implementation artefact of the TCT utilities.
	Reported by Brian Carrier, Purdue University.

	Updated help-recovering-file document by Dan.

Sat Aug 12 10:51:16 EDT 2000

o       Added -V option to direct icat/ils/unrm for verbose output
	to stdout.  The -v option now sends verbose logging to stderr.

Tue Aug  8 11:46:19 PDT 2000

o	created "help-recovering-file", added note to README.FIRST

Sat Aug  5 11:34:26 PDT 2000

o	mactime time range repaired, fixed docs, -l option slightly
	fixed... perhaps deprecation in the future?

Fri Aug  4 20:06:15 EDT 2000

o	ils and ils2mac now include device names for more
	readable mactime reports of dead inodes.

o	mactime could not run with current working directory
	different from $TCT_HOME.

o	mactime -d could destroy an existing $DATA/body file.

o	mactime -d turned on verbosity, instead of letting -v do that.

o	mactime -s (suid) flag used an obsolete method of getting data.

o	ils2mac broke because perl split() can't see the difference
	between empty last fields and missing last fields.

o	mactime mistakenly skipped files with all-zero time stamps.

o	ils2mac now emits ls -l like info for mactime compatibility.

Tue Aug  1 13:50:20 PDT 2000

o	Prepend our directory list to @INC rather than clobbering it.

Mon Jul 31 17:37:35 PDT 2000

o	removed mention of sans.org from help-when-broken-into


extras/entropy/entropy.c
/*++
/* NAME
/*	entropy 1
/* SUMMARY
/*	compute data entropy
/* SYNOPSIS
/* .ad
/* .fi
/*	\fBentropy\fR [\fB-b \fIblock_size\fR] \fIfile...\fR
/* DESCRIPTION
/*	\fBentropy\fR reports entropy values for the named files (standard
/*	input by default). The entropy is computed with Shannon's formula
/*	applied to byte streams. The result gives the average number of bits
/*	of information per byte, under the assumption that the input has no
/*	specific structure.
/*
/*	For each file, entropy is reported as follows:
/*
/* .ti +4
/*	file file_entropy block_count min_entropy max_entropy
/*
/*	Fields are tab separated.
/*
/* .IP \fIfile\fR
/*	The name of the file.
/* .IP \fIfile_entropy\fR
/*	The entropy computed over the entire file contents.
/* .IP \fIblock_count\fR
/*	The number of \fIblock_size\fR blocks used for the minimal and
/*	maximal per-block entropy computation.
/* .IP "\fImin_entropy, max_entropy\fR"
/*	The minimal and maximal per-block entropy for the named file.
/*	No per-block entropy is computed for partial blocks.
/* .PP
/*	 Options:
/* .IP "\fB-b \fIblock_size\fR (default: 8192)"
/*	The block size to be used for per-block entropies.
/* BUGS
/*	Entropy values can be misleading.
/* .IP \(bu
/*	The per-file entropy computed over the entire file is larger
/*	than any per-block entropy value for that same file.
/* .IP \(bu
/*	The entropy values as reported by this program are valid only
/*	for unstructured input. With real data such as text the result
/*	is an overestimate.
/* HISTORY
/* .fi
/* .ad
/*	This \fBentropy\fR command was written for the coroner's toolkit.
/* LICENSE
/*	This software is distributed under the IBM Public License.
/* SEE ALSO
/* .ad
/* .fi
/*      C.E.  Shannon, A mathematical theory of communication. Bell
/*	System Technical Journal 27 (1948) 379-423, 632-656.
/* .nf
/* .na
/*	http://cm.bell-labs.com/cm/ms/what/shannonday/shannon1948.ps.gz
/* AUTHOR(S)
/*	Wietse Venema
/*	IBM T.J. Watson Research
/*	P.O. Box 704
/*	Yorktown Heights, NY 10598, USA
/*--*/

/* System library. */

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <errno.h>
#include <math.h>

/* SunOS4 needs this - I hope it does not break on some pedantic systems. */

extern char *optarg;
extern int optind;

/* Utility library. */

#include <error.h>

/* Application-specific. */

#define ALL_BYTES	256
#define DEF_BLOCK_SIZE	8192

double  log_2;

/* shannon - entropy for a sequence of bytes */

double  shannon(unsigned byte_frequency[ALL_BYTES], unsigned data_length)
{
    double  probability;
    double  sum;
    int     i;

    for (sum = 0, i = 0; i < ALL_BYTES; i++) {
	if (byte_frequency[i]) {
	    probability = (double) byte_frequency[i] / data_length;
	    sum -= probability * log(probability);
	}
    }
    return (sum / log_2);
}

/* process_file - process file one block at a time */

void    process_file(char *file, FILE *fp, int block_length)
{
    unsigned block_frequency[ALL_BYTES];
    unsigned file_frequency[ALL_BYTES];
    unsigned block_count;
    double  max_entropy = 0;
    double  min_entropy = 8;
    double  entropy;
    unsigned i;
    int     ch;

    /*
     * Initialize.
     */
    for (i = 0; i < ALL_BYTES; i++)
	file_frequency[i] = 0;

    /*
     * Process the file, one block at a time.
     */
    for (block_count = 0; /* void */ ; block_count++) {

	/*
	 * Initialize.
	 */
	for (i = 0; i < ALL_BYTES; i++)
	    block_frequency[i] = 0;

	/*
	 * Count character frequencies. Do not report incomplete blocks.
	 */
	for (i = 0; i < block_length; i++) {
	    if ((ch = getc(fp)) == EOF)
		goto done;
	    block_frequency[ch]++;
	    file_frequency[ch]++;
	}

	/*
	 * Compute block entropy.
	 */
	entropy = shannon(block_frequency, block_length);
	if (entropy > max_entropy)
	    max_entropy = entropy;
	if (entropy < min_entropy)
	    min_entropy = entropy;
    }
done:
    if (ferror(fp)) {
	if (errno == EISDIR) {			/* $%@! pedantic Linux */
	    remark("read %s: %m", file);
	    return;
	} else
	    error("read %s: %m", file);
    }

    /*
     * Compute file entropy and report results.
     */
    entropy = shannon(file_frequency, block_count * block_length + i);
    if (block_count == 0)
	min_entropy = max_entropy = 0;
    printf("%s\t%.2f\t%u\t%.2f\t%.2f\n", file, entropy,
	   block_count, min_entropy, max_entropy);
}

/* main - driver */

int     main(int argc, char **argv)
{
    FILE   *fp;
    int     block_length = DEF_BLOCK_SIZE;
    int     ch;
    int     i;

    progname = argv[0];

    /*
     * Parse JCL.
     */
    while ((ch = getopt(argc, argv, "b:")) > 0) {
	switch (ch) {
	case 'b':
	    if ((block_length = atoi(optarg)) <= 0)
		error("bad block size value: %s", optarg);
	    break;
	default:
	    error("usage: %s [-b block length] [file...]", progname);
	}
    }

    /*
     * Sanity checks.
     */

    /*
     * Process files.
     */
    log_2 = log(2.0);

    if (argc == optind) {
	process_file("(stdin)", stdin, block_length);
    } else {
	for (i = optind; i < argc; i++) {
	    if ((fp = fopen(argv[i], "r")) == 0) {
		remark("open %s: %m", argv[i]);
	    } else {
		process_file(argv[i], fp, block_length);
		(void) fclose(fp);
	    }
	}
    }
    exit(0);
}
